modyn/supervisor/internal/eval_strategies/matrix_eval_strategy.py

```diff
@@ -0,0 +1,25 @@
+from typing import Iterable, Optional
+
+from modyn.supervisor.internal.eval_strategies.abstract_eval_strategy import AbstractEvalStrategy
+from modyn.utils import convert_timestr_to_seconds
+
+
+class MatrixEvalStrategy(AbstractEvalStrategy):
+    def __init__(self, eval_strategy_config: dict):
+        super().__init__(eval_strategy_config)
+        self.eval_every = convert_timestr_to_seconds(self.eval_strategy_config["eval_every"])
+        self.eval_start_from = self.eval_strategy_config["eval_start_from"]
+        self.eval_end_at = self.eval_strategy_config["eval_end_at"]
+        assert self.eval_start_from < self.eval_end_at, "eval_start_from must be less than eval_end_at"
+        assert self.eval_every > 0, "eval_every must be greater than 0"
+
+    def get_eval_interval(
+        self, first_timestamp: int, last_timestamp: int
+    ) -> Iterable[tuple[Optional[int], Optional[int]]]:
+        previous_split = self.eval_start_from
+        while True:
+            current_split = min(previous_split + self.eval_every, self.eval_end_at)
+            yield previous_split, current_split
+            if current_split == self.eval_end_at:
```

Comment:
`>=` and `==` has the same effect here because `current_split' is initialized with `eval_start_at` which is asserted to be smaller than `eval_end_at`, and the increment is with a `min`:
`current_split = min(previous_split + self.eval_every, self.eval_end_at)`

In case `eval_end_at` does not align, I stop at `eval_end_at`, i.e. the last interval has length less than `eval_every`.

Reply:
ahh ok it was not obvious to me here that the increment takes care of not stepping over the boundary. anyways, defensive programming is always better :D thanks for adjusting.